test_util/launch.py

```diff
@@ -1,19 +1,139 @@
+
+    def wait(self, info):
+        raise NotImplementedError()
+
+    def describe(self, info):
+        raise NotImplementedError()
+
+    def delete(self, info):
+        raise NotImplementedError()
+
+
+class AwsCloudformationLauncher(AbstractLauncher):
+    def __init__(self, boto_wrapper):
+        self.boto_wrapper = boto_wrapper
+
+    def create(self, config):
+        self.boto_wrapper.create_stack(
+            self.config['stack_name'], self.config['template_url'], self.config['parameters'])
+        return {
+            'type': 'cloudformation',
+            'stack_name': self.config['stack_name'],
+            'provider_info': {
+                'region': self.config['region'],
+                'access_key_id': self.config['access_key_id'],
+                'secret_access_key': self.config['secret_access_key']}}
+
+    def wait(self, info):
+        # TODO: should this support the case where the cluster is being updated?
+        cf = self.get_stack(info)
+        status = cf.get_stack_details()['StackStatus']
+        if status == 'CREATE_IN_PROGRESS':
+            cf.wait_for_stack_creation(wait_for_poll_min=0)
+        elif status == 'CREATE_COMPLETE':
+            pass
+        else:
+            raise Exception('')
+
+    def describe(self, info):
+        cf = self.get_stack(info)
+        return {
+            'masters': convert_host_list(cf.get_master_ips()),
+            'private_agents': convert_host_list(cf.get_private_agent_ips()),
+            'public_agents': convert_host_list(cf.get_public_agent_ips())}
+
+    def delete(self, info):
+        self.get_stack().delete(info)
+
+    def get_stack(self, info):
+        """Returns the correct class interface depending how the AWS CF is configured
+        NOTE: only supports Simple Cloudformation currently
+        """
+        return DcosCfSimple(info['stack_name'], self.boto_wrapper)
+
+
+def get_launcher(launcher_type, provider_info):
+    """Returns a launcher given a launcher type (string) and a dictionary of
+    appropriate provier_info
+    """
+    if launcher_type == 'cloudformation':
+        assert 'region' in provider_info
+        assert 'access_key_id' in provider_info
+        assert 'secret_access_key' in provider_info
+        boto_wrapper = BotoWrapper(
+            provider_info['region'], provider_info['access_key_id'], provider_info['secret_access_key'])
+        return AwsCloudformationLauncher(boto_wrapper)
+    else:
+        raise Exception('Unsupported configuration!')
+
+
+def convert_host_list(host_list):
+    """ Makes Host tuples more readable when using describe
+    """
+    return [{'private_ip': h.private_ip, 'public_ip': h.public_ip} for h in host_list]
 
 
 def main():
-    docopt(__doc__)
+    args = docopt(__doc__, version='DC/OS Launch 1.0')
+
+    if args['create']:
+        info_path = args['--info-path']
+        assert not os.path.exists(info_path), 'There is already a cluster info at {}'.format(info_path)
```

Comment:
shouldn't be an assertion / this is part of main program flow. Assertions are for programmer integrity checking (not meant for user error messages). This should probably be a `if not os.path.exists()` / `print(error)` / `exit(-`)`. Could maybe do an exception + global try/catch/print for the "LauncherException".